crates/wasmtime/src/limiter.rs

```diff
@@ -0,0 +1,129 @@
+use crate::{Store, StoreInner};
+use std::rc::Weak;
+
+/// Used by hosts to limit resource consumption of instances.
+///
+/// A [`Store`] can be created with a resource limiter so that hosts can take into account
+/// non-WebAssembly resource usage to determine if a linear memory or table should grow.
+pub trait ResourceLimiter {
+    /// Notifies the resource limiter that an instance's linear memory has been requested to grow.
+    ///
+    /// * `current` is the current size of the linear memory in WebAssembly page units.
+    /// * `desired` is the desired size of the linear memory in WebAssembly page units.
+    /// * `maximum` is either the linear memory's maximum or a maximum from an instance allocator,
+    ///   also in WebAssembly page units. A value of `None` indicates that the linear memory is
+    ///   unbounded.
+    ///
+    /// This function should return `true` to indicate that the growing operation is permitted or
+    /// `false` if not permitted. Returning `true` when a maximum has been exceeded will have no
+    /// effect as the linear memory will not grow.
+    fn memory_growing(
+        &self,
+        store: &Store,
+        current: u32,
+        desired: u32,
+        maximum: Option<u32>,
+    ) -> bool;
+
+    /// Notifies the resource limiter that an instance's table has been requested to grow.
+    ///
+    /// * `current` is the current number of elements in the table.
+    /// * `desired` is the desired number of elements in the table.
+    /// * `maximum` is either the table's maximum or a maximum from an instance allocator,
+    ///   A value of `None` indicates that the table is unbounded.
+    ///
+    /// This function should return `true` to indicate that the growing operation is permitted or
+    /// `false` if not permitted. Returning `true` when a maximum has been exceeded will have no
+    /// effect as the table will not grow.
+    fn table_growing(
+        &self,
+        store: &Store,
+        current: u32,
+        desired: u32,
+        maximum: Option<u32>,
+    ) -> bool;
+}
+
+pub(crate) struct ResourceLimiterProxy {
+    store: Weak<StoreInner>,
+    limiter: Box<dyn ResourceLimiter>,
+}
+
+impl ResourceLimiterProxy {
+    pub(crate) fn new(store: &Store, limiter: impl ResourceLimiter + 'static) -> Self {
+        Self {
+            store: store.weak(),
+            limiter: Box::new(limiter),
```

Comment:
To avoid double-allocations and double-indirection could `ResourceLimiterProxy` have a type variable perhaps?